Provide more instructions when running git-webkit conflict

akath20 opened the pull request:
#### a790c5b71c6fbfe3350f8ffc99c54b08eb7fe4b3
<pre>
Provide more instructions when running git-webkit conflict
<a href="https://bugs.webkit.org/show_bug.cgi?id=282280">https://bugs.webkit.org/show_bug.cgi?id=282280</a>
<a href="https://rdar.apple.com/132414248">rdar://132414248</a>

Reviewed by Ryan Haddad.

User will now be better informed of what they need to do to resolve the conflict.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
(Conflict):
(Conflict.main):

Canonical link: <a href="https://commits.webkit.org/286165@main">https://commits.webkit.org/286165@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3021fb64ef061d82c9133ebd06e1f66cb7a975c4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75053 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/54484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27873 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/79500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/26304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77170 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63620 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/2269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/79500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/26304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78120 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/49078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/64474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/79500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/74561 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/46423 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/21979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/24630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/67535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/22318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/80976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/2373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/2269 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/80976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/74731 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2523 "Build is in progress. Recent messages:") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/64491 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/80976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/16519 "The change is no longer eligible for processing.") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/8581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11575 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/2338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/2366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/3292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/2373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

bravc commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -84,11 +84,31 @@ def main(cls, args, repository, **kwargs):
+
+        source_changes = radar_obj.source_changes
+        if len(source_changes) != 1:
+            print('Checked out conflict branch with conflict markers present. Please fix, amend and force push to resolve.')
```
So this only works for single source radars? It's not clear from this message that is the issue

bravc commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -84,11 +84,32 @@ def main(cls, args, repository, **kwargs):
+        msg += f'\n\ngit reset --hard {source_remote}/{conflict_pr.base}\n'
+        for source in radar_obj.source_changes:
+            source_sha = source.split(', ')[2]
+            msg += f'git cherry-pick {source_sha}\n'
```
I think f-string is a no-go in this codebase but I could be wrong

gsnedders replied:
f-strings should be fine since 5200ed9 when we started mandating Python 3.6

bravc commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -84,11 +84,32 @@ def main(cls, args, repository, **kwargs):
+        msg += "Please resolve them, amend the commit and force push.\n"
+        msg += "\nAlternatively, if you want to get into the traditional conflict state (ex. `git status` shows conflicting files)\n"
+        msg += "you can run the following commands. Note: This will remove any changes that were edited on the branch directly."
+        msg += f'\n\ngit reset --hard {source_remote}/{conflict_pr.base}\n'
```
Still an f-string here

briannafan replied:
I think f-strings won't break things but format strings are preferred

gsnedders replied:
As I said in the resolved thread, f-strings should be fine since 5200ed9 when we started mandating Python 3.6

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -84,11 +84,32 @@ def main(cls, args, repository, **kwargs):
+        msg += "Conflict markers are present in files.\n"
+        msg += "Please resolve them, amend the commit and force push.\n"
+        msg += "\nAlternatively, if you want to get into the traditional conflict state (ex. `git status` shows conflicting files)\n"
+        msg += "you can run the following commands. Note: This will remove any changes that were edited on the branch directly."
```
Are "changes that were edited on the branch directly", equivalent to "any working changes"? That might be less confusing for the user.  Also a nit, but maybe "Warning" is better than "Note" since `git reset --hard` is scary!

briannafan replied:
Oh I re-read this and do you mean any commits on the branch will be lost since we're resetting to the base?

akath20 replied:
Thanks for calling this out! That is true, but so will the remove pull request changes (if another author edits lets say).

I talked with @bravc to revise the wording here.

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
```diff
@@ -84,11 +84,32 @@ def main(cls, args, repository, **kwargs):
+        msg += "Conflict markers are present in files.\n"
+        msg += "Please resolve them, amend the commit and force push.\n"
+        msg += "\nAlternatively, if you want to get into the traditional conflict state (ex. `git status` shows conflicting files)\n"
+        msg += "you can run the following commands. Warning: This will require a force push. Any changes made to the pull request branch will therefore be lost."
```
Final nit: Capitalize `you`!

akath20 replied:
if you want to get into the traditional conflict state (ex. `git status` shows conflicting files) you can run the following commands"

is the full line, even though it's broken up, it's not a new sentence

briannafan replied:
Ohhh

akath20 committed:
Provide more instructions when running git-webkit conflict
https://bugs.webkit.org/show_bug.cgi?id=282280
rdar://132414248

Reviewed by Ryan Haddad.

User will now be better informed of what they need to do to resolve the conflict.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/conflict.py:
(Conflict):
(Conflict.main):

Canonical link: https://commits.webkit.org/286165@main
